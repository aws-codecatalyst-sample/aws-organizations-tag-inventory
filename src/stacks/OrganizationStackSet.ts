/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */


import {CfnStackSet, Stack, StackProps, Tags} from 'aws-cdk-lib';

import {NagSuppressions} from 'cdk-nag';
import {Construct} from 'constructs';
import * as cdk from "aws-cdk-lib/core/lib";



export interface OrganizationStackSetProps extends StackProps {
	templateUrl:string,
	stackInstancesGroup:Array<cdk.IResolvable | CfnStackSet.StackInstancesProperty> | cdk.IResolvable
}

export class OrganizationStackSet extends Stack {

	constructor(scope: Construct, id: string, props: OrganizationStackSetProps) {
		super(scope, id, props);
		new CfnStackSet(this,"aws-organizations-tag-inventory-spoke-account-stack-set",{
			description: "StackSet for deploy the aws-organizations-tag-inventory-spoke-stack to account across the organization",
			stackSetName: "aws-organizations-tag-inventory-spoke-account-stack-set",
			permissionModel: "SERVICE_MANAGED",
			capabilities:["CAPABILITY_IAM"],
			autoDeployment: {
				enabled: false,
				retainStacksOnAccountRemoval:false
			},
			templateUrl: props.templateUrl,
			stackInstancesGroup: props.stackInstancesGroup
		})
		this.cdkNagSuppressions();
		Tags.of(this).add('Solution', 'aws-organizations-tag-inventory');
		Tags.of(this).add('Url', 'https://github.com/aws-samples/aws-organizations-tag-inventory');
	}

	private cdkNagSuppressions() {

		NagSuppressions.addStackSuppressions(this, []);
	}
}