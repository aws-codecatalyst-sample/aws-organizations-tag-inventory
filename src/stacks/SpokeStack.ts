import path from 'path';
import { Aws, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Effect, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Architecture, LayerVersion, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { CfnSchedule } from 'aws-cdk-lib/aws-scheduler';
import { Construct } from 'constructs';
import { Layers } from '../constructs/Layers';
import { ResourceExplorerIndex } from '../constructs/ResourceExplorerIndex';
import { StateMachineFromFile } from '../constructs/StateMachineFromFile';

export interface SpokeStackProps extends StackProps {
  enabledRegions: string[];
  aggregatorRegion: string;
  bucketName: string | undefined;
  centralRoleArn: string | undefined;

}

export class SpokeStack extends Stack {

  constructor(scope: Construct, id: string, props: SpokeStackProps) {
    super(scope, id, props);
    if (!props.bucketName) {
      throw new Error('Bucket name is required');

    }
    if (!props.centralRoleArn) {
      throw new Error('Central role ARN is required');

    }
    const powerToolsLayer = LayerVersion.fromLayerVersionArn(this, 'powertools', `arn:aws:lambda:${Aws.REGION}:094274105915:layer:AWSLambdaPowertoolsTypeScript:11`);

    const layers = new Layers(this, 'layers');
    //put resources here
    const resourceExplorerIndex = new ResourceExplorerIndex(this, 'MyIndex', {
      layers: layers,
      enabledRegions: props.enabledRegions,
      aggregatorRegion: props.aggregatorRegion,

    });


    const searchFunction = new NodejsFunction(this, 'Search-fn', {
      architecture: Architecture.ARM_64,
      runtime: Runtime.NODEJS_18_X,
      entry: path.join(__dirname, '..', 'functions', 'Search.ts'),
      handler: 'index.onEvent',
      timeout: Duration.seconds(60),
      layers: [layers.layer, powerToolsLayer],
      initialPolicy: [new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['resource-explorer-2:Search'],
        resources: ['*'],
      })],
      environment: {
        VIEW_ARN: resourceExplorerIndex.viewArn,
        LOG_LEVEL: 'DEBUG',
      },
    });
    const mergeFunction = new NodejsFunction(this, 'MergeResults-fn', {
      architecture: Architecture.ARM_64,
      runtime: Runtime.NODEJS_18_X,
      entry: path.join(__dirname, '..', 'functions', 'MergeResults.ts'),
      handler: 'index.onEvent',
      timeout: Duration.seconds(60),
      layers: [powerToolsLayer],
      initialPolicy: [],
      environment: {
        LOG_LEVEL: 'DEBUG',
      },
    });

    const stateMachine = new StateMachineFromFile(this, 'SpokeAccountStateMachine',
      { name: 'SpokeAccountStateMachine', file: path.join(__dirname, '..', 'stateMachines', 'SpokeAccountStateMachine.json'), searchFunction: searchFunction, mergeFunction: mergeFunction, putObjectRoleArn: props.centralRoleArn, bucketName: props.bucketName });
    stateMachine.stateMachine.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['dynamodb:PutItem', 'dynamodb:BatchWriteItem', 'dynamodb:UpdateItem'],
      resources: ['*'],


    }));
    stateMachine.stateMachine.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['s3:PutObject', 's3:PutObjectAcl'],
      resources: [`arn:aws:s3:::${props.bucketName}`, `arn:aws:s3:::${props.bucketName}/*`],
    }));
    stateMachine.stateMachine.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['sts:AssumeRole'],
      resources: [props.centralRoleArn],
    }));
    mergeFunction.grantInvoke(stateMachine.stateMachine);
    searchFunction.grantInvoke(stateMachine.stateMachine);
    const role=new Role(this, 'SchedulerRole', { assumedBy: new ServicePrincipal('scheduler.amazonaws.com') });
    stateMachine.stateMachine.grantStartExecution(role);
    new CfnSchedule(this, 'Scheduler', {
      name: 'TagInventorySchedule',
      flexibleTimeWindow: {
        maximumWindowInMinutes: 60,
        mode: 'FLEXIBLE',
      },
      state: 'ENABLED',
      scheduleExpression: 'rate(1 days)',
      target: {
        arn: stateMachine.stateMachine.stateMachineArn,
        roleArn: role.roleArn,
      },
      scheduleExpressionTimezone: 'America/New_York',
    });

  }
}