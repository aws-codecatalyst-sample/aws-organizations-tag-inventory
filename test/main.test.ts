import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SpokeStack } from '../src/stacks/SpokeStack';

test('Snapshot', () => {
  const app = new App();
  const stack = new SpokeStack(app, 'test', {
    enabledRegions: ['us-east-1', 'us-east-2', 'us-west-1', 'us-west-2'],
    aggregatorRegion: 'us-east-2',
    bucketName: 'test-bucket',
    centralRoleArn: 'test-role',
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});