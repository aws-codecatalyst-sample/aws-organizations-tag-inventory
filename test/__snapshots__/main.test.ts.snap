// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Outputs": {
    "MyIndextaginventoryallresourcesarnoutputFF1B3C48": {
      "Description": "Arn of tag-inventory-all-resources view",
      "Value": {
        "Fn::GetAtt": [
          "MyIndexresourceexplorerindexcustomresource6FC13F1F",
          "ViewArn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyIndexresourceexplorerindexcustomresource6FC13F1F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AGGREGATOR_INDEX_REGION": "us-east-2",
        "ENABLED_REGIONS": [
          "us-east-1",
          "us-east-2",
          "us-west-1",
          "us-west-2",
        ],
        "ServiceToken": {
          "Fn::GetAtt": [
            "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventCA9D2DDB",
            "Arn",
          ],
        },
      },
      "Type": "Custom::ResourceExplorerIndex",
      "UpdateReplacePolicy": "Delete",
    },
    "Searchfn39BBF692": {
      "DependsOn": [
        "SearchfnServiceRoleDefaultPolicy46A0B0D1",
        "SearchfnServiceRole3456AD27",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4f27b8111ac1f6e4f9f3d36bc2228dfe79c7fa9cf424e99c0c032d749b85d5b2.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "LOG_LEVEL": "DEBUG",
            "VIEW_ARN": {
              "Fn::GetAtt": [
                "MyIndexresourceexplorerindexcustomresource6FC13F1F",
                "ViewArn",
              ],
            },
          },
        },
        "Handler": "index.onEvent",
        "Layers": [
          {
            "Ref": "layerslayerB10D4ECB",
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:lambda:",
                {
                  "Ref": "AWS::Region",
                },
                ":094274105915:layer:AWSLambdaPowertoolsTypeScript:11",
              ],
            ],
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "SearchfnServiceRole3456AD27",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SearchfnServiceRole3456AD27": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SearchfnServiceRoleDefaultPolicy46A0B0D1": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "resource-explorer-2:Search",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SearchfnServiceRoleDefaultPolicy46A0B0D1",
        "Roles": [
          {
            "Ref": "SearchfnServiceRole3456AD27",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SpokeAccountStateMachine10567581": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "SpokeAccountStateMachineRoleDefaultPolicyF9A051B5",
        "SpokeAccountStateMachineRole9B579215",
      ],
      "Properties": {
        "DefinitionString": "{
  "Comment": "A description of my state machine",
  "StartAt": "Initialize Next Token",
  "States": {
    "Initialize Next Token": {
      "Type": "Pass",
      "Next": "Search",
      "Result": {
        "NextToken": null
      }
    },
    "Search": {
      "Type": "Task",
      "Resource": "arn:aws:states:::lambda:invoke.waitForTaskToken",
      "OutputPath": "$.Payload",
      "Parameters": {
        "Payload.$": "$",
        "FunctionName": "\${SEARCH_FUNCTION}"
      },
      "Retry": [
        {
          "ErrorEquals": [
            "Lambda.ServiceException",
            "Lambda.AWSLambdaException",
            "Lambda.SdkClientException",
            "Lambda.TooManyRequestsException"
          ],
          "IntervalSeconds": 2,
          "MaxAttempts": 6,
          "BackoffRate": 2
        }
      ],
      "Next": "Map"
    },
    "Map": {
      "Type": "Map",
      "ItemProcessor": {
        "ProcessorConfig": {
          "Mode": "INLINE"
        },
        "StartAt": "Pass",
        "States": {
          "Pass": {
            "Type": "Pass",
            "OutputPath": "$.Arn",
            "End": true
          }
        }
      },
      "ItemsPath": "$.Resources",
      "ResultPath": null,
      "Next": "Are there more results?"
    },
    "Are there more results?": {
      "Type": "Choice",
      "Choices": [
        {
          "Variable": "$.NextToken",
          "IsPresent": true,
          "Comment": "NextToken : Present",
          "Next": "Search"
        }
      ],
      "Default": "Success"
    },
    "Success": {
      "Type": "Succeed"
    }
  }
}",
        "DefinitionSubstitutions": {
          "SEARCH_FUNCTION": {
            "Fn::GetAtt": [
              "Searchfn39BBF692",
              "Arn",
            ],
          },
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "SpokeAccountStateMachineRole9B579215",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "SpokeAccountStateMachineRole9B579215": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "SpokeAccountStateMachineRoleDefaultPolicyF9A051B5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Searchfn39BBF692",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Searchfn39BBF692",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SpokeAccountStateMachineRoleDefaultPolicyF9A051B5",
        "Roles": [
          {
            "Ref": "SpokeAccountStateMachineRole9B579215",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventCA9D2DDB": {
      "DependsOn": [
        "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRoleDefaultPolicy196222C7",
        "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRole6629EB69",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "8e3d635893ea17fa3158623489cd42c680fad925b38de1ef51cb10d84f6e245e.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (test/com.amazonaws.cdk.custom-resources.resource-explorer-index-provider/resource-explorer-index-provider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnE4EA2129",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRole6629EB69",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRole6629EB69": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRoleDefaultPolicy196222C7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnE4EA2129",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnE4EA2129",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRoleDefaultPolicy196222C7",
        "Roles": [
          {
            "Ref": "comamazonawscdkcustomresourcesresourceexplorerindexproviderframeworkonEventServiceRole6629EB69",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnE4EA2129": {
      "DependsOn": [
        "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRoleDefaultPolicy74BEE050",
        "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRole2F22B68A",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b8e091a77bcc50bb4224e76797cf692cbdb2669fd2e6e0a8434d54ffbd9dfeb0.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.onEvent",
        "Layers": [
          {
            "Ref": "layerslayerB10D4ECB",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRole2F22B68A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRole2F22B68A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRoleDefaultPolicy74BEE050": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "resource-explorer-2:CreateView",
                "resource-explorer-2:ListViews",
                "resource-explorer-2:GetView",
                "resource-explorer-2:GetIndex",
                "resource-explorer-2:CreateIndex",
                "resource-explorer-2:UpdateIndexType",
                "resource-explorer-2:GetDefaultView",
                "resource-explorer-2:AssociateDefaultView",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "iam:CreateServiceLinkedRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::*:role/aws-service-role/resource-explorer-2.amazonaws.com/AWSServiceRoleForResourceExplorer",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRoleDefaultPolicy74BEE050",
        "Roles": [
          {
            "Ref": "comamazonawscdkcustomresourcesresourceexplorerindexproviderresourceexplorerindexfnServiceRole2F22B68A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "layerslayerB10D4ECB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "CompatibleArchitectures": [
          "arm64",
          "x86_64",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2e1211f8b940b0f9f163f1d0024547982cd4cd1a113bcec9926727fd203bf9f5.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
